pkg/controller/jobframework/setup.go

```diff
@@ -51,57 +51,71 @@ var (
 // until the webhooks are operating, and the webhook won't work until the
 // certs are all in place.
 func SetupControllers(ctx context.Context, mgr ctrl.Manager, log logr.Logger, opts ...Option) error {
-	return manager.setupControllers(ctx, mgr, log, opts...)
+	discoveredCRDs := make(chan schema.GroupVersionKind, 10)
+	go watchCRDs(ctx, mgr, log, discoveredCRDs)
+	return manager.setupControllersFromDiscoveredCRDs(ctx, mgr, log, discoveredCRDs, opts...)
 }
 
-func (m *integrationManager) setupControllers(ctx context.Context, mgr ctrl.Manager, log logr.Logger, opts ...Option) error {
+func (m *integrationManager) setupControllersFromDiscoveredCRDs(ctx context.Context, mgr ctrl.Manager, log logr.Logger, discoveredCRDs <-chan schema.GroupVersionKind, opts ...Option) error {
 	options := ProcessOptions(opts...)
 
-	if err := m.checkEnabledListDependencies(options.EnabledFrameworks); err != nil {
-		return fmt.Errorf("check enabled frameworks list: %w", err)
-	}
-
-	for fwkName := range options.EnabledExternalFrameworks {
-		if err := RegisterExternalJobType(fwkName); err != nil {
-			return err
-		}
-	}
-	return m.forEach(func(name string, cb IntegrationCallbacks) error {
-		logger := log.WithValues("jobFrameworkName", name)
-		fwkNamePrefix := fmt.Sprintf("jobFrameworkName %q", name)
-
-		if options.EnabledFrameworks.Has(name) {
-			if cb.CanSupportIntegration != nil {
-				if canSupport, err := cb.CanSupportIntegration(opts...); !canSupport || err != nil {
-					log.Error(err, "Failed to configure reconcilers")
-					os.Exit(1)
+	setupChan := make(chan struct{}, 10)
+
+	go func() {
+		for gvk := range discoveredCRDs {
+			setupChan <- struct{}{}
+			go func(currentGVK schema.GroupVersionKind) {
+				defer func() { <-setupChan }()
+
+				var matchedName string
+				var matchedCallbacks *IntegrationCallbacks
+				m.mu.Lock()
+				for name, cb := range m.integrations {
+					if options.EnabledFrameworks.Has(name) {
+						if name == fmt.Sprintf("%s/%s", currentGVK.Group, strings.ToLower(currentGVK.Kind)) ||
+							strings.EqualFold(reflect.TypeOf(cb.JobType).String(), currentGVK.Kind) {
```

Comment:
The constant use of `reflect` could cause a slower runtime performance. I'd suggest to use pre-computed strings for direct string comparison.